text/0000-partial-template.md

```diff
@@ -0,0 +1,101 @@
+---
+title: Partial Template
+status: DRAFTED
+created_at: 2020-01-16
+updated_at: 2020-01-17
+pr: (leave this empty until the PR is created)
+---
+
+# Short Title
+
+## Summary
+
+Partial templates allow the ability to import html templates into another template via
+the normal variable syntax `{template}`.
+
+## Basic example
+
+```js
+import partialTemplate from './partialTemplate.html';
+
+export default class Example extends LightningElement {
+    partialTemplate = partialTemplate;
+}
+```
+
+```html
+<template>
+  Parent
+  {partialTemplate}
+</template>
+```
+
+## Motivation
+
+In investigating performance issues in components that are used hundreds of times on a
+page it was noted that creating reusable child components; while organizationally nice,
+were not as performant as inlining the HTML into the template;
+
+In other instances it was noted that large conditonal blocks of HTML could be more organized
+into seperate template files.
+
+## Detailed design

```

Comment:
The biggest question for me here is what is the scope of the template? can it be controlled? or we assume that the template have access to the component instance? To be clear, I'm talking about `{x}` inside the partial, what does it do? is that equivalent to `{x}` in the importer?

Also, I think the fact that you need JS to use this makes me doubt that this is the correct approach.

Reply:
Would 100% assume it would work the same as if it the partial's content had been pasted in.

```html
template = 
<template>Hello {name}!</template>

class { template = importedTemplate; name = 'world'; }
<template>{template}</template>

Would result in:
Hello world!
```

Same would be the case anywhere the `{template}` is placed picking up any scoped variables.

Need to look into how templates are processed now to better understand how this could work. The alternative being templates are compile time placed in, but that limits what they can do and would be mainly for splitting up templates code.

Reply:
well, this is faulty for two main reasons:

1. it is not a reusable partial because you don't have a way to specify what should be accessible to it.
2. it is impossible for the compiler (at its current state) to invoke another function (compiled partial) while providing the same JS scope of the caller.

While 2 is just a technical limitation that we can attempt to figure, I believe 1 is an ergonomic issue that is a deal breaker here, let me put an example:

```html
// foo.html
<template><p>Value: {value}</p></template>

// component.html
<template>
      <h1>Before</h1>
      {foo}
      <h2>After</h1>
      {foo}
</template>
```

I'm trying to reuse a piece of the UI so I don't have to duplicate it in my template, but also I don't have to worry about the two fragments to go out of sync. This is clearly a use-case for partials, yet it doesn't work because the partial expects `value` to be accessible, but I have two values in the caller that I will like to display in two different places. That's why I think the scope MUST be specified, e.g.:

```html
// foo.html
<template><p>Value: {value}</p></template>

// component.html
<template>
      <h1>Before</h1>
      <template lwc:partial="./foo" value={myBeforeValue}></template>
      <h2>After</h1>
      <template lwc:partial="./foo" value={myAfterValue}></template>
</template>
```

It also provides some declarative only syntax where the value of the `lwc:partial` directive doesn't have to be an imported value in JS if the partial is relative.

Reply:
Updated the alternative section with the example above.

That greatly simplifies the usage and allows more flexibility with template reuse. Assuming one could pair that with `if:true`/`if:false`.

Would there be a huge jump in complexity between that and...
```html
<template lwc:partial={template} value={val}></template>
```

Reply:
Would that work for events as well? Let's assume that my partial has a button and I'd like to trigger an action onclick. How can I assign an action? Is it something like this:
```html
<template lwc:partial="./foo" onbuttonclick={myHandler}></template>

<template>
  <button onclick={onbuttonclick}>Mybutton</button>
<template>
```

Reply:
@priandsf Assuming when it compiles down it has a way to wire up the scope it probably could do the same for events. Not really familiar enough with how the internals work though (just assuming it's a compile time in this case).

(this proposal was put on hiatus, but still something that would be beneficial)